### PR TITLE
[flutter_tools] support screenshot on all device types

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fallback to VM Service "rasterizer" screenshot type and/or chrome VM Service screenshot type.

Actually wire up web screenshot functionality.

Move testing to terminal handler and delete more heavily mocked resident_runner tests

### CODE REVIEW

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 1]
```diff
@@ -938,10 +939,16 @@ abstract class ResidentHandlers {
   /// If the device has a connected vmservice, this method will attempt to hide
   /// and restore the debug banner before taking the screenshot.
   ///
-  /// Throws an [AssertionError] if [Device.supportsScreenshot] is not true.
+  /// If the device type does not support a "native" screenshot, then this
+  /// will fallback to a rasterizer screenshot from the engine. This has the
+  /// downside of being unable to display the contents of platform views.
+  ///
+  /// Currently the behavior of screenshot is to bail if it encounters any
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/resident_runner.dart:946]: What does "bail" mean here? Throw a tool exit? Silently fail? Fail with a printError? etc.

Maybe also rephrase without "Currently", like "Fails with an error message and continues if there is an RPC error".

File: packages/flutter_tools/test/general.shard/terminal_handler_test.dart [hunk 2]
```diff
@@ -953,6 +948,210 @@ void main() {
...
+      supportsScreenshot: false,
+      fileSystem: fileSystem,
+    );
+
+    await terminalHandler.processTerminalInput('s');
+
+    expect(logger.errorText, contains('Error'));
+  });
+
+  testWithoutContext('s, bails taking screenshot on debug device if debugAllowBanner during second request', () async {
```
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/terminal_handler_test.dart:1122]: What happens if the vmservice screenshot throws an RPC error?

Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/terminal_handler_test.dart:1195]: We attempt to restore the debug banner. I added 2 more test cases

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/resident_runner.dart]: Took another pass on this

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 3]
```diff
@@ -938,10 +939,18 @@ abstract class ResidentHandlers {
   /// If the device has a connected vmservice, this method will attempt to hide
   /// and restore the debug banner before taking the screenshot.
   ///
-  /// Throws an [AssertionError] if [Device.supportsScreenshot] is not true.
+  /// If the device type does not support a "native" screenshot, then this
+  /// will fallback to a rasterizer screenshot from the engine. This has the
+  /// downside of being unable to display the contents of platform views.
+  ///
+  /// This method wil return without writing the screenshot file if any
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/resident_runner.dart:946]:
```suggestion
  /// This method will return without writing the screenshot file if any
```

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 4]
```diff
@@ -938,10 +939,18 @@ abstract class ResidentHandlers {
...
+  /// RPC errors are encountered, printing them to stderr. This is true even
+  /// if an error occurs after the data has already been recieved, such as
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/resident_runner.dart:948]:
```suggestion
  /// if an error occurs after the data has already been received, such as
```

**Review by zanderso — APPROVED** [review]

**Comment by jonahwilliams** [comment]
This is what I get for using the online editor without spellcheck configured, thanks @zanderso  😓

**Pull request merged**